src/server.rs

```diff
@@ -0,0 +1,409 @@
+        let session: &ServerSession = try_ref_from_ptr!(session, &ServerSession, false);
+        session.wants_write()
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn rustls_server_session_is_handshaking(
+    session: *const rustls_server_session,
+) -> bool {
+    ffi_panic_boundary_bool! {
+        let session: &ServerSession = try_ref_from_ptr!(session, &ServerSession, false);
+        session.is_handshaking()
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn rustls_server_session_process_new_packets(
+    session: *mut rustls_server_session,
+) -> rustls_result {
+    ffi_panic_boundary! {
+        let session: &mut ServerSession = try_ref_from_ptr!(session, &mut ServerSession);
+        match session.process_new_packets() {
+            Ok(()) => rustls_result::Ok,
+            Err(e) => return map_error(e),
+        }
+    }
+}
+
+/// Queues a close_notify fatal alert to be sent in the next write_tls call.
+/// https://docs.rs/rustls/0.19.0/rustls/trait.Session.html#tymethod.send_close_notify
+#[no_mangle]
+pub extern "C" fn rustls_server_session_send_close_notify(session: *mut rustls_server_session) {
+    ffi_panic_boundary_unit! {
+        let session: &mut ServerSession = try_ref_from_ptr!(session, &mut ServerSession, ());
+        session.send_close_notify()
+    }
+}
+
+/// Free a server_session previously returned from rustls_server_session_new.
+/// Calling with NULL is fine. Must not be called twice with the same value.
+#[no_mangle]
+pub extern "C" fn rustls_server_session_free(session: *mut rustls_server_session) {
+    ffi_panic_boundary_unit! {
+        let session: &mut ServerSession = try_ref_from_ptr!(session, &mut ServerSession, ());
+        // Convert the pointer to a Box and drop it.
+        unsafe { Box::from_raw(session); }
+    }
+}
+
+/// Write up to `count` plaintext bytes from `buf` into the ServerSession.
+/// This will increase the number of output bytes available to
+/// `rustls_server_session_write_tls`.
+/// On success, store the number of bytes actually written in *out_n
+/// (this may be less than `count`).
+/// https://docs.rs/rustls/0.19.0/rustls/struct.ServerSession.html#method.write
+#[no_mangle]
+pub extern "C" fn rustls_server_session_write(
+    session: *mut rustls_server_session,
+    buf: *const u8,
+    count: size_t,
+    out_n: *mut size_t,
+) -> rustls_result {
+    ffi_panic_boundary! {
+        let session: &mut ServerSession = try_ref_from_ptr!(session, &mut ServerSession);
+        let write_buf: &[u8] = unsafe {
+            if buf.is_null() {
+                return NullParameter;
+            }
+            slice::from_raw_parts(buf, count as usize)
+        };
+        let out_n: &mut size_t = unsafe {
+            match out_n.as_mut() {
+                Some(out_n) => out_n,
+                None => return NullParameter,
+            }
+        };
+        let n_written: usize = match session.write(write_buf) {
+            Ok(n) => n,
+            Err(_) => return rustls_result::Io,
```

Comment:
Kind of a shame to drop whatever information might be in the error returned from `session.write()`. I know you have thought a ton about error handling so I apologize if you've already made up your mind about this but couldn't `rustls_result` contain an arbitrary error string?

Reply:
I definitely considered some way of returning errors with arbitrary strings. It would be nice in various ways! But there are some annoying downsides. For instance, say we turn `rustls_result` into a `struct { int code; size_t msglen; const char* msg }`. We have to choose one of "crustls owns the memory, and your rustls_result becomes invalid after any subsequent rustls call" (doesn't work for concurrent programs) or "caller takes ownership of the memory and has to remember to free it after getting an error" (easy for caller to leak memory; not leaking memory requires possibly-awkward `goto cleanup`). I'm interested in learning about best C practices here. From what I can tell, it seems like most libraries rely on enum-valued errors.

By the way, I think in the specific `Io` error cases in this library we don't need to worry. We're not doing real I/O, we're writing to a session's in-memory buffer. Internally this is infallible: https://docs.rs/rustls/0.19.0/src/rustls/client/mod.rs.html#787-800.